### ISSUE TITLE
$ref pointing to subdirectory generates classes in wrong packages

### ISSUE THREAD

**Issue by rfoltyns** [joelittlejohn/jsonschema2pojo#410]
When $ref includes a schema from different directory, referred class is generated in the package containing referring schema, e.g.:

*schema/bar/a.json*
```javascript
{
  "$schema" : "http://json-schema.org/draft-04/schema#",
  "type" : "object",
  "properties" : {
    "refA" : {
      "$ref": "../foo/b.json"
    }
  }
}
```

*schema/foo/b.json*
```javascript
{
  "$schema" : "http://json-schema.org/draft-04/schema#",
  "type" : "object",
  "properties" : {
    "propA" : {
      "type" : "string"
    }
  }
}
```

Both classes are generated in *bar* package.

I assumed that:
- sourceDirectory is the default package of generated classes 
- directory structure along with the schemas inside will be reflected in generated packages and classes

It's mainly caused by a this block of code in SchemaRule class:
```java
        if (schemaNode.has("$ref")) {
            schema = ruleFactory.getSchemaStore().create(schema, schemaNode.get("$ref").asText());
            schemaNode = schema.getContent();

            if (schema.isGenerated()) {
                return schema.getJavaType();
            }

            return apply(nodeName, schemaNode, generatableType, schema); // <-- see description below
        }
```

When *generatableType* is a *JPackage*, it points to current package instead of package of referred schema.

I have a piece of code that fixes this issue:
- generates classes in proper directories when $ref points to subdirectory (at any level)
- handles $ref containing #
- unit tests for common and edge cases, covering 100% of added/modified code
- leaves other protocols (http, java, resource, file, etc.) intact

But the questions is: Is it an bug or an improvement?
If it's an improvement, we should discuss the compliance with the spec of your plugin and handling of other protocols if needed.
If it's a bug, here's the patch fix necessary changes and tests:
https://github.com/joelittlejohn/jsonschema2pojo/compare/master...badhabit85:patch-1

**Comment by ctrimble** [comment]
@badhabit85 this seems like a long standing bug that people have been using `javaType` to get around.  I took a quick look at your patch and it seems strange that you have actual protocol names in there.

**Comment by badhabit85** [comment]
According to the tests I've found [here](https://github.com/joelittlejohn/jsonschema2pojo/tree/master/jsonschema2pojo-integration-tests/src/test/java/org/jsonschema2pojo/integration/ref) plugin can use the $ref value to include schema from many different sources (e.g.: classpath, web, etc.). The patch resolves the bug for relative paths only and doesn't impact the others as I didn't know how they should be treated - for instance: should the value of 'classpath:schema/foo/bar/A.json' be resolved to 'foo.bar' package or 'schema.foo.bar'? 
The protocols' names you mentioned are in the **exclusion list** - they're not a part of the package resolution process for 2 purposes:
1. don't break the contract of handling other protocols (most important)
2. leave the plugin open for further improvements (e.g.: handle http, classpath, resource using protocol-specific approach)
'javaType' is a workaround, but I believe that we can't workaround forever, hence the fix.

**Comment by abazylewicz** [comment]
+1

**Comment by dadrus** [comment]
I've included "support" for ids (referenced in #481), which can solve the issue. Would you take a look on it?

**Comment by badhabit85** [comment]
It works like a charm. Thank you!

**Comment by adityasharat** [comment]
Any updates on this issue ?

**Comment by rfoltyns** [comment]
I've just checked it using 0.4.27. Looks fine, thank you :)

**Comment by sergeykad** [comment]
The issues may have been solved for Maven plugin in  #481, but not for others. Classes are still generated in wrong packages if using Gradle.
